sources/host-ctr/cmd/host-ctr/pid_namespaces.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// Get file descriptor that refers to a parent namespace, ioctl() request '_IO(NSIO, 0x2)'
+	nsGetParentNs        = 0xb702
+	lockfile      string = "/run/lock/host-ctr-pid-ns.lock"
+	manifestPath  string = "/etc/host-containers/info/pid-namespaces.json"
+)
+
+func getTaskPidNs(task *containerd.Task) (int64, error) {
+	pid := (*task).Pid()
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	_, err = unix.IoctlGetInt(pidNsFd, nsGetParentNs)
+	if err == syscall.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb syscall.Stat_t
+	if err := syscall.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+func addPidNsToManifest(pidNs int64) error {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	locked, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	if !locked {
+		return errors.New("failed to acquire lock for writing to PID NS manifest")
+	}
+	defer func() {
+		_ = fileLock.Unlock()
+	}()
+
+	// Add task PID NS to manifest
+	if err := os.MkdirAll(filepath.Dir(manifestPath), 0666); err != nil {
+		return errors.Wrap(err, "failed to create PID NS manifest directory")
+	}
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
```

Comment:
Typically, instead of allocating memory ourselves, would be to call `os.ReadFile` which returns a byte slice anyways.

```go
func os.ReadFile(name string) ([]byte, error)
```

Reply:
`os.ReadFile` opens the file and reads everything then closes the file.
I don't want to `syscall.Open` the file again when I have to `os.OpenFile` before this with custom flags and modes, I just want to do everything with the single `manifest` open file descriptor.

Reply:
Yeah I was wondering if that was the strategy you were going for since you already call `Stat` on the file descriptor 👏🏼 that works! Maybe a comment noting why we need the open file descriptor?

Reply:
> Maybe a comment noting why we need the open file descriptor?

Not sure what I'd say to clarify what's happening since I believe the flow of what's happening to the `fd` seems pretty straightforward.

Reply:
👏🏼  sounds good!